### PR TITLE
API-213: extract attribute and locale checks

### PR DESCRIPTION
**Pull request by solivier**
[//]: <> (<3 Thanks for taking the time to contribute! You're awesome! <3)

[//]: <> (If you've never contributed to this repository before, please read https://github.com/akeneo/pim-community-dev/blob/master/.github/CONTRIBUTING.md)

**Description (for Contributor and Core Developer)**

[//]: <> (What does this Pull Request do? reference the related issue?)

Tests will be added later

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Yes
| Added Behats                      | No
| Added integration tests           | Yes
| Changelog updated                 | No
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo

### CODE REVIEW

File: src/Pim/Bundle/ApiBundle/Checker/ProductQueryParametersCheckerInterface.php [hunk 1]
```diff
@@ -0,0 +1,44 @@
+<?php
+
+namespace Pim\Bundle\ApiBundle\Checker;
+
+use Pim\Component\Catalog\Model\ChannelInterface;
+use Symfony\Component\HttpKernel\Exception\UnprocessableEntityHttpException;
+
```
Comment (skeleton) [inline src/Pim/Bundle/ApiBundle/Checker/ProductQueryParametersCheckerInterface.php]: useless

File: src/Pim/Bundle/ApiBundle/Checker/ProductQueryParametersCheckerInterface.php [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+/**
```
Comment (skeleton) [inline src/Pim/Bundle/ApiBundle/Checker/ProductQueryParametersCheckerInterface.php]: add a description please

File: src/Pim/Bundle/ApiBundle/Controller/ProductController.php [hunk 3]
```diff
@@ -102,31 +103,34 @@ class ProductController
     /** @var array */
     protected $apiConfiguration;
 
+    /** @var ProductQueryParametersCheckerInterface */
+    protected $productQueryParametersChecker;
```
Comment (skeleton) [inline src/Pim/Bundle/ApiBundle/Controller/ProductController.php]: remove property $localeRepository

File: src/Pim/Bundle/ApiBundle/Checker/ProductQueryParametersChecker.php [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+/**
+ * @author    Olivier Soulet <olivier.soulet@akeneo.com>
+ * @copyright 2017 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
+ */
+class ProductQueryParametersChecker implements ProductQueryParametersCheckerInterface
+{
+    /** @var IdentifiableObjectRepositoryInterface */
+    protected $localeRepository;
```
Comment (skeleton) [inline src/Pim/Bundle/ApiBundle/Checker/ProductQueryParametersChecker.php]: private

File: src/Pim/Bundle/ApiBundle/Checker/ProductQueryParametersChecker.php [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+    /** @var IdentifiableObjectRepositoryInterface */
```
Comment (skeleton) [inline src/Pim/Bundle/ApiBundle/Checker/ProductQueryParametersChecker.php]: same

File: src/Pim/Bundle/ApiBundle/Checker/ProductQueryParametersChecker.php [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    ) {
+        $this->localeRepository = $localeRepository;
+        $this->attributeRepository = $attributeRepository;
+        $this->categoryRepository = $categoryRepository;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function checkLocalesParameters($localeCodes, ChannelInterface $channel = null)
```
Comment (skeleton) [inline src/Pim/Bundle/ApiBundle/Checker/ProductQueryParametersChecker.php]: array $localeCodes (and you explode data in the controller) ? by this way, if at one moment, the separator is not a comma, we won't be blocked.
Same comment for all method in this class

File: src/Pim/Bundle/ApiBundle/Checker/ProductQueryParametersCheckerInterface.php (hunk 1 above)
Comment (solivier) [inline src/Pim/Bundle/ApiBundle/Checker/ProductQueryParametersCheckerInterface.php]: what is useless ?

File: src/Pim/Bundle/ApiBundle/Checker/ProductQueryParametersChecker.php [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+use Akeneo\Component\StorageUtils\Repository\IdentifiableObjectRepositoryInterface;
+use Pim\Component\Catalog\Model\ChannelInterface;
+use Symfony\Component\HttpKernel\Exception\UnprocessableEntityHttpException;
+
+/**
+ * @author    Olivier Soulet <olivier.soulet@akeneo.com>
+ * @copyright 2017 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
+ */
+class ProductQueryParametersChecker implements ProductQueryParametersCheckerInterface
```
Comment (skeleton) [inline src/Pim/Bundle/ApiBundle/Checker/ProductQueryParametersChecker.php]: QueryParamterChecker is enough imho, nothing is related to the product here

File: src/Pim/Bundle/ApiBundle/Checker/ProductQueryParametersChecker.php (hunk 6 above)
Comment (solivier) [inline src/Pim/Bundle/ApiBundle/Checker/ProductQueryParametersChecker.php]: good idea I'll do this

File: src/Pim/Bundle/ApiBundle/Checker/ProductQueryParametersChecker.php (hunk 7 above)
Comment (solivier) [inline src/Pim/Bundle/ApiBundle/Checker/ProductQueryParametersChecker.php]: right I'll change

File: src/Pim/Bundle/ApiBundle/Checker/ProductQueryParametersCheckerInterface.php (hunk 1 above)
Comment (skeleton) [inline src/Pim/Bundle/ApiBundle/Checker/ProductQueryParametersCheckerInterface.php]: double blank line :)

File: src/Pim/Bundle/ApiBundle/Controller/ProductController.php [hunk 8]
```diff
@@ -145,7 +146,7 @@ public function __construct(
         $this->pqbFactory = $pqbFactory;
         $this->normalizer = $normalizer;
         $this->channelRepository = $channelRepository;
-        $this->localeRepository = $localeRepository;
```
Comment (skeleton) [inline src/Pim/Bundle/ApiBundle/Controller/ProductController.php]: don't forget to add BC break in changelog

Comment (solivier) [inline src/Pim/Bundle/ApiBundle/Controller/ProductController.php]: done

**Pull request merged**